handle SqlrecoverableException

Lorren opened the pull request:
Moved and combined from #994  #1032 #1039  to new pull request. 

Handles SQLRecoverableException  using generic interface method parameters and properly close dboject  connections as per documentation. 

Please review.

Lorren committed:
Handle SQLRecoverableException. Expose remove abandoned properties

Lorren committed:
fix spaces and log messages

Lorren committed:
additional sqlrecoverabelexception handlers, empty property checking

Lorren committed:
additional sqlrecoverabelexception handlers, empty property checking

Lorren committed:
Merge remote-tracking branch 'origin/sqlrecoverable-4.4.x' into sqlrecoverable-4.4.x

Conflicts:
	core/org.wso2.carbon.user.core/src/main/java/org/wso2/carbon/user/core/util/DatabaseUtil.java

Lorren committed:
Merge remote-tracking branch 'origin/sqlrecoverable-4.4.x' into sqlrecoverable-4.4.x

Lorren committed:
Merge remote-tracking branch 'origin/sqlrecoverable-4.4.x' into sqlrecoverable-4.4.x

Lorren committed:
Merge remote-tracking branch 'origin/sqlrecoverable-4.4.x' into sqlrecoverable-4.4.x

Lorren committed:
Merge remote-tracking branch 'origin/sqlrecoverable-4.4.x' into sqlrecoverable-4.4.x

Lorren committed:
Use try-with-resources

Lorren committed:
merge sqlrecoverable-4.4.x

Conflicts:
	core/org.wso2.carbon.user.core/src/main/java/org/wso2/carbon/user/core/util/DatabaseUtil.java

Lorren committed:
use interface in handleSQLRecoverable

Lorren committed:
use interface in handlesqlrecoverable method call.  add unit test to repository

Lorren committed:
merge sqlrecoverable-4.4.x --> 4.4.x

Lorren committed:
Java7 multi-catch syntax . use generic argument types for to use interfaces to remove code duplication.

Lorren committed:
merge sqlrecoverable-4.4.x --> 4.4.x

Lorren committed:
Java7 multi-catch syntax . use generic argument types for to use interfaces to remove code duplication.

Lorren committed:
Java7 multi-catch syntax . use generic argument types for to use interfaces to remove code duplication.

Lorren committed:
Java7 multi-catch syntax . use generic argument types for to use interfaces to remove code duplication.

Lorren committed:
merge changes sqlrecoverable-4.4.x --> 4.4.x

Lorren committed:
prepare handleSqlrecoverable for merge

imesh commented:
Similar to your first PR, the new PR seems to have large blocks of NEW and DELETED code. As a result, we cannot see the modified lines clearly. It would be better if you can resend the PR with proper MODIFIED, NEW and DELETED lines. To do this you might need to avoid formatting code or update your IDE to match with the exiting formatting of the source files.

Can you please also explain the steps you followed to verify this functionality?
Thanks

Lorren commented:
Interesting, it doesn’t look like that on my fork:
https://github.com/Lorren/carbon-kernel/commit/6c6885b1b9d5eec303f67e2475a1a20b69ab36e0

From: Imesh Guaratne [mailto:notifications@github.com]
Sent: Thursday, June 30, 2016 8:53 AM
To: wso2/carbon-kernel carbon-kernel@noreply.github.com
Cc: Kay, Lorren (Plymouth Meeting 2) LKay@us.imshealth.com; Author author@noreply.github.com
Subject: Re: [wso2/carbon-kernel] handle SqlrecoverableException (#1068)

As I mentioned to you in your first PR, the new PR seems to have large blocks of NEW and DELETED code. As a result, we cannot see the modified lines clearly. It would be better if you can resend the PR proper MODIFIED, NEW and DELETED lines. To do this you might need to avoid formatting code or update your IDE to match with the exiting formatting of the source files.

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHubhttps://github.com/wso2/carbon-kernel/pull/1068#issuecomment-229649106, or mute the threadhttps://github.com/notifications/unsubscribe/AD9dJJgkeieNrwg3myXgUvUJZyRrrGplks5qQ7xAgaJpZM4I-GcD.

Lorren committed:
Merge remote-tracking branch 'upstream/4.4.x' into 4.4.x

Lorren committed:
Merge branch '4.4.x' of https://github.com/lorren/carbon-kernel into 4.4.x

Lorren committed:
Merge branch '4.4.x' of https://github.com/lorren/carbon-kernel into 4.4.x

Lorren committed:
Merge branch 'sqlrecoverable-4.4.x' of https://github.com/lorren/carbon-kernel into sqlrecoverable-4.4.x

Lorren commented:
Rebased my fork, you can see the changes now.  
The unit tests confirm functionality. 
Thanks

imesh commented:
Thanks Lorren! Now the diff is clear. Will review it soon.

kasunbg commented on core/org.wso2.carbon.user.core/src/main/java/org/wso2/carbon/user/core/util/DatabaseUtil.java:
```diff
@@ -27,11 +27,7 @@
-import java.sql.ResultSet;
-import java.sql.SQLException;
-import java.sql.Timestamp;
+import java.sql.*;
```
Star imports are discouraged as a convention at WSO2. Please try to expand. Thanks.

Lorren replied:
Done, Thanks.

ramith commented on core/org.wso2.carbon.user.core/src/main/java/org/wso2/carbon/user/core/util/DatabaseUtil.java:
```diff
@@ -571,71 +572,100 @@ public static Connection getDBConnection(DataSource dataSource) throws SQLExcept
             try {
-                rs.close();
+                dbObject.close();
+            } catch (SQLRecoverableException ex) {
```
Does jtds , and microsoft drivers throw this exception (I suppose not). if thats the case you might have to resort to sqlstate class code. https://docs.oracle.com/javase/7/docs/api/java/sql/SQLException.html#getSQLState()

Lorren replied:
Thank you for your review.   There is no need to change the code.  If the exception isn't thrown, the catch statement will not be entered, one of the other catch statements will handle it.  SQLException "should" be  thrown by every driver out there.  In the off-chance it doesn't, then every exception inherits from the Exception class.

https://docs.oracle.com/javase/tutorial/jdbc/basics/sqlexception.html#categorized_sqlexceptions

imesh commented on core/org.wso2.carbon.user.core/src/main/java/org/wso2/carbon/user/core/util/DatabaseUtil.java:
```diff
@@ -571,71 +572,100 @@ public static Connection getDBConnection(DataSource dataSource) throws SQLExcept
-            } catch (SQLException e) {
-                log.error("Database error. Could not close statement. Continuing with others. - " + e.getMessage(), e);
-            }
+    private static <AutoClosableWrapper extends AutoCloseable & Wrapper>  void handleSQLRecoverableException(AutoClosableWrapper dbObject, SQLRecoverableException recException){
```
I think it might be better if this method name can imply that it is only handling SQLRecoverableException of databaseObject.close() operation. On the other hand I do not see a reason to retry a close operation because at the time close operation is invoked the database operation may have already completed.

Lorren replied:
This can be used to handle any sqlrecoverableexcepiton, performing minimal actions specified in the API.

Lorren replied:
Changed the name.

imesh commented:
@Lorren As I noticed you have only handled SQLRecoverableException on databaseObject.close() operation. I think we would also need to handle this on following methods in JDBCUserStoreManager:

```
prepStmt.executeQuery()
prepStmt.executeUpdate()
getDBConnection()
```

imesh commented on core/org.wso2.carbon.user.core/src/test/java/org/wso2/carbon/user/core/util/DatabaseUtilTest.java:
```diff
@@ -0,0 +1,265 @@
+            Mockito.doThrow(new SQLRecoverableException("throw SQLRecoverableException")).when(conn).close();
+            DatabaseUtil.closeConnection(conn);
+            System.out.println("Expected error thrown in sqlerecoverable handler");
+            Assert.assertNotNull(conn);
```
Would you mind explaining how this assertion works? AFAIU, database connection object would never be null irrespective of the actual database connection status. Thanks!

Lorren replied:
Removed the vestigial test code.

imesh commented on core/org.wso2.carbon.user.core/src/test/java/org/wso2/carbon/user/core/util/DatabaseUtilTest.java:
```diff
@@ -0,0 +1,265 @@
+            DatabaseUtil.close(conn, preparedStatement);
+        } catch (SQLRecoverableException e) {
+            isThrown = true;
+            Assert.assertTrue(isThrown);
```
Wouldn't this assertion be always true?

imesh commented on core/org.wso2.carbon.user.core/src/test/java/org/wso2/carbon/user/core/util/DatabaseUtilTest.java:
```diff
@@ -0,0 +1,265 @@
+            // this shouldn't happen
+            Assert.assertTrue(false);
+        }
+        Assert.assertNotNull(preparedStatement);
```
Would you mind explain how this assertion works? Thanks!

imesh commented on core/org.wso2.carbon.user.core/src/test/java/org/wso2/carbon/user/core/util/DatabaseUtilTest.java:
```diff
@@ -0,0 +1,265 @@
+        } catch (SQLRecoverableException e) {
+            // this is expected
+            isThrown = true;
+           Assert.assertTrue(isThrown);
```
Similar to the above scenario, this assertion will be always true as the line before is setting the variable isThrown to true.

imesh commented on core/org.wso2.carbon.user.core/src/test/java/org/wso2/carbon/user/core/util/DatabaseUtilTest.java:
```diff
@@ -0,0 +1,265 @@
+           Assert.assertTrue(isThrown);
+
+        } catch (SQLException e) {
+            e.printStackTrace();
```
It would be better to log exceptions than using e.printStackTrace():
http://stackoverflow.com/questions/10477607/avoid-printstacktrace-use-a-logger-call-instead

Lorren replied:
Cleaned all the tests up  Thanks.

Lorren committed:
explicitly declare sql  imports.  refactor and rename unit tests

Lorren committed:
explicitly declare sql  imports.  refactor and rename unit tests

Lorren commented:
@Imesh AFIK the exception hasn’t been thrown from any of those calls.   But if you think you need it, you can follow the same model I created in this class.

From: Imesh Guaratne [mailto:notifications@github.com]
Sent: Tuesday, July 12, 2016 8:42 AM
To: wso2/carbon-kernel carbon-kernel@noreply.github.com
Cc: Kay, Lorren (Plymouth Meeting 2) LKay@us.imshealth.com; Mention mention@noreply.github.com
Subject: Re: [wso2/carbon-kernel] handle SqlrecoverableException (#1068)

@Lorrenhttps://github.com/Lorren As I noticed you have only handled SQLRecoverableException for databaseObject.close() operation. I think we would also need to handle this for following methods in JDBCUserStoreManager:

prepStmt.executeQuery()

prepStmt.executeUpdate()

getDBConnection()

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHubhttps://github.com/wso2/carbon-kernel/pull/1068#issuecomment-232035398, or mute the threadhttps://github.com/notifications/unsubscribe/AD9dJM7yxgCQPqQNPg5l6UV0UrdOPgrTks5qU4ungaJpZM4I-GcD.

imesh commented on core/org.wso2.carbon.user.core/src/test/java/org/wso2/carbon/user/core/util/DatabaseUtilTest.java:
```diff
@@ -0,0 +1,260 @@
+        Mockito.doThrow(new SQLRecoverableException("throw SQLRecoverableException")).when(resultSetMocker.get()).next();
+
+        String[] results =   DatabaseUtil.getStringValuesFromDatabase(conn, sqlStmt, params.get(0), params.get(1), params.get(2));
+        Assert.fail("Can not handle SQLRecoverableError" );
```
Typo: Can not => Cannot

Lorren replied:
@imesh Not really a typo. It was intended for emphasis.  But, you didn't like it, so I changed it..

Lorren committed:
change  text per comments.

imesh commented:
Thanks for doing the refinements Lorren! We will review this again with the Carbon Kernel team next week and merge it. Thanks again!

kasunbg commented on core/org.wso2.carbon.user.core/src/main/java/org/wso2/carbon/user/core/jdbc/JDBCRealmConstants.java:
```diff
@@ -145,6 +145,9 @@
     public static final String MIN_EVIC_TABLE_IDLE_TIME_MILLIS = "minEvictableIdleTimeMillis";
     public static final String NUM_TESTS_PEREVICTION_RUN = "numTestsPerEvictionRun";
+    public static final String REMOVE_ABANDONED = "removeAbandoned";
+    public static final String REMOVE_ABANDNONED_TIMEOUT = "removeAbandonedTimeout";
```
minor typo in the variable name "ABANDNONED".

darshanasbg commented on core/org.wso2.carbon.user.api/src/main/java/org/wso2/carbon/user/api/RealmConfiguration.java:
```diff
@@ -97,7 +100,29 @@ public boolean isPasswordsExternallyManaged() {
         this.passwordsExternallyManaged = passwordsExternallyManaged;
     }
-
+    public boolean isRemoveAbandoned(){
```
Can\Need to avoid API changes in RealmConfiguration

darshanasbg commented on core/org.wso2.carbon.user.core/src/main/java/org/wso2/carbon/user/core/util/DatabaseUtil.java:
```diff
@@ -466,19 +479,14 @@ public static void udpateUserRoleMappingWithExactParams(Connection dbConnection,
-                DatabaseUtil.closeAllConnections(dbConnection);
-            }
-            DatabaseUtil.closeAllConnections(null, ps);
+            close(dbConnection);
```
Since the connection getting passed as a parameter to this method, it shouldn't close the connection inside this method.

Need to close the connection in the original place where it originally (create and )close the dbConnection.

darshanasbg commented on core/org.wso2.carbon.user.core/src/main/java/org/wso2/carbon/user/core/util/DatabaseUtil.java:
```diff
@@ -557,7 +563,7 @@ public static void updateDatabase(Connection dbConnection, String sqlStmt, Objec
             throw new UserStoreException(errorMessage, e);
         } finally {
-            DatabaseUtil.closeAllConnections(null, prepStmt);
+            close(dbConnection);
```
Since the connection getting passed as a parameter to this method, it shouldn't close the connection inside this method.

Need to close the connection in the original place where it originally (create and )close the dbConnection.

darshanasbg commented on core/org.wso2.carbon.user.core/src/main/java/org/wso2/carbon/user/core/util/DatabaseUtil.java:
```diff
@@ -262,35 +282,35 @@ private static DataSource createRealmDataSource(RealmConfiguration realmConfig)
+
         } finally {
-            DatabaseUtil.closeAllConnections(null, rs, prepStmt);
+            close(dbConnection);
```
Since the connection getting passed as a parameter to this method, it shouldn't close the connection inside this method.

Need to close the connection in the original place where it originally (create and )close the dbConnection.

darshanasbg commented on core/org.wso2.carbon.user.core/src/main/java/org/wso2/carbon/user/core/util/DatabaseUtil.java:
```diff
@@ -303,38 +323,37 @@ private static DataSource createRealmDataSource(RealmConfiguration realmConfig)
-        } finally {
-            DatabaseUtil.closeAllConnections(null, rs, prepStmt);
+        }finally {
+            close(dbConnection);
```
Since the connection getting passed as a parameter to this method, it shouldn't close the connection inside this method.

Need to close the connection in the original place where it originally (create and )close the dbConnection.

darshanasbg commented on core/org.wso2.carbon.user.core/src/main/java/org/wso2/carbon/user/core/util/DatabaseUtil.java:
```diff
@@ -347,28 +366,27 @@ public static int getIntegerValueFromDatabase(Connection dbConnection, String sq
-        } finally {
-            DatabaseUtil.closeAllConnections(null, rs, prepStmt);
+        }finally {
+            close(dbConnection);
```
Since the connection getting passed as a parameter to this method, it shouldn't close the connection inside this method.

Need to close the connection in the original place where it originally (create and )close the dbConnection.

darshanasbg commented on core/org.wso2.carbon.user.core/src/main/java/org/wso2/carbon/user/core/util/DatabaseUtil.java:
```diff
@@ -419,22 +437,17 @@ public static void udpateUserRoleMappingInBatchModeForInternalRoles(Connection d
-            }
-            DatabaseUtil.closeAllConnections(null, prepStmt);
+        } finally {  // can remove since this is not needed with try-with-resources -
+            close(dbConnection);
```
Since the connection getting passed as a parameter to this method, it shouldn't close the connection inside this method.

Need to close the connection only `localConnection` set to true.

Lorren committed:
made change per comments, exposed properties, closed dbconnectiosn,

Lorren committed:
made change per comments, exposed properties, closed dbconnectiosn,

JKAUSHALYA commented:
Thanks @Lorren for the PR. We have created another PR (#1091) on top of this PR including the fix you gave for the issue [CARBON-16012](https://wso2.org/jira/browse/CARBON-16012) to avoid API changes. Since we need to handle the SQL Recoverable Exceptions properly in a common manner for the platform, we'll include that change later.

Lorren commented:
The SQL Recoverable Exception Handling was the impetus for the code changes.  The subset contains only the additional properties that provide debugging information for the SQLRecoverableException.

The pull also excludes the bug fix in closeAllConnections .

From: Jayanga Kaushalya [mailto:notifications@github.com]
Sent: Wednesday, August 10, 2016 5:02 AM
To: wso2/carbon-kernel carbon-kernel@noreply.github.com
Cc: Kay, Lorren (Plymouth Meeting 2) LKay@us.imshealth.com; Mention mention@noreply.github.com
Subject: Re: [wso2/carbon-kernel] handle SqlrecoverableException (#1068)

A sub set of changes has been added to the PR #1091https://github.com/wso2/carbon-kernel/pull/1091 to avoid API changes. As we have some uncertainties regarding handling the SQL Recoverable Exception, it is also not included.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHubhttps://github.com/wso2/carbon-kernel/pull/1068#issuecomment-238807409, or mute the threadhttps://github.com/notifications/unsubscribe-auth/AD9dJK2KTB-toUHCxEP0xhwVxZ67zz20ks5qeZOQgaJpZM4I-GcD.

Lorren commented:
What uncertainties do you have regarding handling the SQL Recoverable Exception?

JKAUSHALYA commented:
@Lorren As I mentioned in my updated comment, we need to handle the recoverable exception in a common manner, if we are going to handle that.

In your [handleSQLRcoverableExceptionOnClose](https://github.com/wso2/carbon-kernel/pull/1068/files#diff-6bb5fa840eb13b5bd934a72646d332fcR603) method, what you are doing is closing the closable object which has already thrown an exception when trying to close. So it may throw another exception (Same exception). And then, you are getting a new connection and closing it. That might not be the expected procedure of handling a recoverable exception [1].

For example, you are only closing the connection, statement, etc.. after a recoverable exception. But we may need to re execute or rollback the entire operation depending on the place where it occurred after getting a new connection ([Without closing it right after](https://github.com/wso2/carbon-kernel/pull/1068/files#diff-6bb5fa840eb13b5bd934a72646d332fcR607)).

And there are some other reasons also. For example as in [here](https://github.com/wso2/carbon-kernel/pull/1068/files#diff-6bb5fa840eb13b5bd934a72646d332fcR304), we can't close the connection as it is passed as a parameter and it should be closed at that level. (Which is mentioned in the inline comment as well)

[1] https://docs.oracle.com/javase/7/docs/api/java/sql/SQLRecoverableException.html

Lorren committed:
Fix whitespace for merge branch 'sqlrecoverable-4.4.x' of https://github.com/lorren/carbon-kernel into sqlrecoverable-4.4.x

Lorren commented:
I've updated the code per the comments.

I understand wanting consistent handling for the exception.  However, these really must be handled not merely logged.  We are seeing a number of stale connections returning to the pool, taking up slots.     When a new request for a connection comes the stale connections are being served out.  Where the connections are being closed is unknown.  Closing these explicitly disposes of them gracefully so bad connections are invalidated and valid connections can be established.  If you want to handle these consistently do so, but please don't ignore them.

JKAUSHALYA commented:
Closing this since we have another PR (https://github.com/wso2/carbon-kernel/pull/1091).
